addons/website_sale/models/website.py

```diff
@@ -60,35 +60,53 @@ def _get_pl_partner_order(self, country_code, show_visible, website_pl, current_
         :param int order_pl: the current cart pricelist
         :returns: list of pricelist ids
         """
+        def _check_show_visible(pl):
+            """ If `show_visible` is True, we will only show the pricelist if
+            one of this condition is met:
+            - The pricelist is `selectable`.
+            - The pricelist is either the currently used pricelist or the
+            current cart pricelist, we should consider it as available even if
+            it might not be website compliant (eg: it is not selectable anymore,
+            it is a backend pricelist, it is not active anymore..).
+            """
+            return (not show_visible or pl.selectable or pl.id in (current_pl, order_pl))
+
+        # Note: 1. pricelists from all_pl are already website compliant (went through `_get_website_pricelists_domain`)
+        #       2. do not read `property_product_pricelist` here as this method is cached and this a change of this field
+        #          does not invalidate the cache. Pass it through `partner_pl` parameter instead.
+
         self.ensure_one()
         pricelists = self.env['product.pricelist']
         if country_code:
             for cgroup in self.env['res.country.group'].search([('country_ids.code', '=', country_code)]):
-                for group_pricelists in cgroup.pricelist_ids.filtered(lambda pl: pl.website_id.id in (False, self.id)):
-                    if not show_visible or group_pricelists.selectable or group_pricelists.id in (current_pl, order_pl):
-                        pricelists |= group_pricelists
+                pricelists |= cgroup.pricelist_ids.filtered(lambda pl: pl._is_available_on_website(self.id) and _check_show_visible(pl))
 
-        partner = self.env.user.partner_id.sudo(user=self.env.user)
         is_public = self.user_id.id == self.env.user.id
-        if not is_public and (not pricelists or (partner_pl or partner.property_product_pricelist.id) != website_pl):
-            # `property_product_pricelist` is already multi-website compliant
-            pricelists |= partner.property_product_pricelist
-
-        if not pricelists:  # no pricelist for this country, or no GeoIP
-            pricelists |= all_pl.filtered(lambda pl: not show_visible or pl.selectable or pl.id in (current_pl, order_pl))
-        if not show_visible and not country_code:
-            pricelists |= all_pl.filtered(lambda pl: pl.sudo().code)
+        # If the user is logged in and either he has no country pricelists or either
+        # his partner pricelist is not the website pricelist, then we use his partner pricelist
+        if not is_public and (not pricelists or partner_pl != website_pl):
+            pricelists |= pricelists.browse(partner_pl)
+
+        if is_public and not pricelists:  # no pricelist for this country, or no GeoIP
+            # TODO (review) if is_public was implicit as you cant go in this is if you are not public given the previous if
```

Comment:
What?

Reply:
Knowing that `partner_pl` is the `property_product_pricelist` of the partner, that field is now already 'website compliant' with this PR (in a frontend context).

The previous if is:
```python
if not is_public and (not pricelists or partner_pl != website_pl):`
    pricelists |= pricelists.browse(partner_pl)
```

So after that condition,
```python
if not pricelists:
```
Is the same as
```python
if is_public and not pricelists:
```
As the previous if, if you are not public and you don't have a pricelist will add `partner_pl` in the `pricelists` variable. So if you don't have a pricelist at that time, you are by deduction public.

Unless I am missing something :yum: